hazelcast/src/main/java/com/hazelcast/config/ConfigXmlGenerator.java

```diff
@@ -986,9 +987,11 @@ private static void mapConfigXmlGenerator(XmlGenerator gen, Config config) {
                     .node("metadata-policy", m.getMetadataPolicy());
 
             evictionConfigXmlGenerator(gen, m.getEvictionConfig());
-            appendMerkleTreeConfig(gen, m.getMerkleTreeConfig());
+            if (m.getMerkleTreeConfig().getEnabled() != null) {
+                appendMerkleTreeConfig(gen, m.getMerkleTreeConfig());
+            }
             appendEventJournalConfig(gen, m.getEventJournalConfig());
-            appendHotRestartConfig(gen, m.getHotRestartConfig());
+            appendDataPersistenceConfig(gen, m.getDataPersistenceConfig());
```

Comment:
instead of deprecated hot-restart config, we should generate the current data-persistence config

Reply:
Shouldn't we cover both configs to keep the generated configs equality? (Config1->XML->Config2)

Reply:
Since we are merging `HotRestartPersistenceConfig` and `PersistenceConfig` during transformation from/to XML, a config with `HotRestartPersistence` -> XML (output as the new `persistence` element) -> parsed as config2 with `PersistenceConfig` will be equal to the source config. See also https://github.com/hazelcast/hazelcast/pull/19004#issuecomment-874478305. Is my reasoning correct @fbarotov ?

Reply:
hmm, yeah we merge both `HotRestartPersistenceConfig` and `PersistenceConfig` when reading from XML, but when writing to XML we should write only `HotRestartPersistenceConfig` if the target can be HZ with version less than 5.0, otherwise just `DataPersistenceConfig` would do, since as Vassilis mentioned we merge two configs every time one changes.

Reply:
On equality before/after transformation to/from XML, I just pushed a commit to fix it. I thought HotRestartPersistence-Persistence configs merge was already there but it was actually missing in `ConfigXmlGenerator#generate` method. Thanks for catching that @kwart .

On whether we should output `hot-restart-persistence` or `persistence` element: I suppose that starting with 5.0 we want people to migrate from the deprecated `hot-restart-persistence` configuration elements to the new `persistence` configuration. That's why I think we should output `persistence` in the generated XML from 5.0.
If we were to output both elements, then `hot-restart-persistence` would be a partial copy of the `persistence` element, since any new config will be only added to `persistence` config (eg see https://github.com/hazelcast/hazelcast/pull/19522/files#diff-ff0ffd59a12a51408de4fdac0b50ff4634ee86dad470c63d69a17359a815d16eR4049).
wdyt?

Reply:
I discussed the issue with @fbarotov. In my opinion, we shouldn't proactively change the Config object. We should keep the config which user has configured. And the same config also store/generate to the XML. We should only resolve the proper persistence settings on places where we use it.

Reply:
Faruk will work on a related issue hazelcast/hazelcast-enterprise#4281

Reply:
Without doing a deep-dive on the consequences of this config merging, I'd say let's just keep it for now, unless there are severe correctness issues, rolling upgrade issues or something along those lines.

IMHO the conundrum of config equality when using `ConfigXmlGenerator` is not something we should dwell on for too long, especially this close to the release.

In other words, let's keep the current mechanism of proactively changing the config, until someone starts complaining. Maybe nobody will complain, and we'll remove the HR config in (hopefully) 5.1.

Reply:
> On whether we should output hot-restart-persistence or persistence element: I suppose that starting with 5.0 we want people to migrate from the deprecated hot-restart-persistence configuration elements to the new persistence configuration. That's why I think we should output persistence in the generated XML from 5.0.

Agree with this.